src/main/java/org/wildfly/transaction/client/provider/jboss/JBossXAResourceRegistry.java

```
@@ -0,0 +1,275 @@
+            }
+            // remove resource from in doubt list, in case the resource was in doubt
+            inDoubtResources.remove(resource);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void resourceInDoubt(XAResource resource) {
+        inDoubtResources.add(resource);
+    }
+
+    /**
+     * Loads in doubt resources from recovered registry file.
+     *
+     * @throws IOException if an I/O error occurs when reloading the resources from the file
+     */
+    private void loadInDoubtResources(String nodeName) throws IOException {
+        assert fileChannel == null;
+        final List<String> uris;
+        try {
+            uris = Files.readAllLines(filePath);
+        } catch (IOException e) {
+            throw Log.log.readXAResourceRecoveryFileFailed(filePath, e);
+        }
+        for (String uriString: uris) {
+            final URI uri;
+            try {
+                uri = new URI(uriString);
+            } catch (URISyntaxException e) {
+                throw Log.log.readURIFromXAResourceRecoveryFileFailed(uriString, filePath, e);
+            }
+            final XAResource xaresource = reloadInDoubtResource(uri, nodeName);
+            inDoubtResources.add(xaresource);
+            Log.log.xaResourceRecoveredFromRecoveryRegistry(uri, filePath);
+        }
+    }
+
+    /**
+     * Returns XAResourceRecovery for in doubt XAResources recovery.
+     *
+     * @return the XAResourceRecovery for in doubt subordinate XA resources recovery
+     * @see org.jboss.tm.XAResourceRecoveryRegistry#addXAResourceRecovery(XAResourceRecovery)
+     */
+    static XAResourceRecovery getXAResourceRecovery(LocalTransactionProvider provider) {
+        return () -> {
+            try {
+                recoverInDoubtRegistries(provider);
+            } catch (IOException e) {
+                // ignore with a logged message
+                Log.log.unexpectedExceptionOnXAResourceRecovery(e);
```

I think this should probably throw an exception, like an ``IllegalStateException`` perhaps which wraps the ``IOException``